### PR TITLE
feat: add thaumcraft integration to search for essentia

### LINKED ISSUE #17591: Let FindIt find essentia within Jars
### Your GTNH Discord Username

leumasme

### Your Pack Version

2.6.1

### Your Proposal

When using FindIt "Find Item in nearby Inventories" on an Aspect from the Thaumcraft NEI Plugin, nearby Thaumcraft containers (Jars, Alchemical Furnace, Arcane Alembic, ...) containing that Aspect/Essentia should be highlighted.

This aligns with the existing goal of FindIt, since it is already able to find Fluids within containers and pipes.

Possibly make it also consider Phial of Essentia as the same thing when searching/finding to mirror the behavior of Fluids + Fluid Container Items?

This shouldn't be too hard to implement as Waila is already able to query the Essentia contained within the previously mentioned Containers (only shown when wearing Goggles of Revealing)

### Your Goal

This specific change makes it easier to find Essentia when you have a lot of jars.

### Your Vision

Integrate FindIt with all mods where it makes sense to make it work universally.

### Final Checklist

- [X] I have searched this issue tracker and there is nothing similar already. Posting on a closed issue saying *I like this feature please reconsider adding it* will prompt us to investigate and reopen it once we confirm your report.
- [X] I believe there is nothing similar in the pack already, or the existing solution isn't good enough.
- [X] I understand this change request may not attract enough attention and thus not be implemented.
- [X] I understand this change request may be rejected due to other community members thinking it's inappropriate.
- [X] I believe this feature would make the pack better.

### PR DESCRIPTION
**Pull request by leumasme**
Closes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/17591

https://github.com/user-attachments/assets/6b4239ef-f8d4-40b9-a1a7-612c9322b5c1

Searching for an Aspect Item from Thaumcraft NEI Integration will find Items (Phials, Jars) and Blocks (Nodes, Jars, ...) that contain essentia of that aspect.

This currently only happens when doing the Find Item on the Aspect Item specifically. It could easily be added that this also happens when pressing it on a filled Phial or Jar directly, if that's desirable (would mirror the behavior for fluid ex. water bucket).

### CODE REVIEW

**Comment by leumasme** [comment]
The way that the Filter system since #22 is structured is a bit inverted for this (and in general?).
A filter is created for the TileEntity / ItemStack that is currently being inspected to check if it's a matching one, not the one that we're searching for.
This way, each filter is only being used once (Can't search for multiple Items at once + Filter isn't cached (not that it should)).
This means that (if necessary) extracting information from the ItemStack we're searching for has to be done *again for every TileEntity/ItemStack* that is getting checked. Here, that extra work is 'just' getting the Aspect of the Aspect Item, but in the future if there is a case that needs more work done on the item being searched, this design may be very inefficient,
so it seems like an odd choice.

File: src/main/java/com/gtnh/findit/FindIt.java [hunk 1]
```diff
@@ -4,12 +4,7 @@
 import java.util.List;
 
 import com.gtnh.findit.IStackFilter.IStackFilterProvider;
-import com.gtnh.findit.handler.AdventureBackpackProvider;
-import com.gtnh.findit.handler.BackpackProvider;
-import com.gtnh.findit.handler.DraconicEvolutionProvider;
-import com.gtnh.findit.handler.ForestryStackFilterProvider;
-import com.gtnh.findit.handler.MinecraftProvider;
-import com.gtnh.findit.handler.ProjectRedExplorationProvider;
+import com.gtnh.findit.handler.*;
```
Comment (slprime) [inline src/main/java/com/gtnh/findit/FindIt.java]: don't use `*` in import.

**Comment by slprime** [comment]
> The way that the Filter system since #22 is structured is a bit inverted for this (and in general?). A filter is created for the TileEntity / ItemStack that is currently being inspected to check if it's a matching one, not the one that we're searching for. 
This way, each filter is only being used once (Can't search for multiple Items at once + Filter isn't cached (not that it should)). This means that (if necessary) extracting information from the ItemStack we're searching for has to be done _again for every TileEntity/ItemStack_ that is getting checked. Here, that extra work is 'just' getting the Aspect of the Aspect Item, but in the future if there is a case that needs more work done on the item being searched, this design may be very inefficient, so it seems like an odd choice.

Correct. this is necessary because each block can have its own algorithm for reading its inventory.
It is impossible to build a filter from the side of the ItemStack you are looking for, since the filter is unique for each TileEntity/ItemStack  in which you need to search (If you have an idea how to fix this, give an example)

**Comment by slprime** [comment]
> This currently only happens when doing the Find Item on the Aspect Item specifically. It could easily be added that this also happens when pressing it on a filled Phial or Jar directly, if that's desirable (would mirror the behavior for fluid ex. water bucket).

I don't have an answer to that. I don't play with magic and I don't know how it would be convenient.

**Comment by YannickMG** [comment]
> > This currently only happens when doing the Find Item on the Aspect Item specifically. It could easily be added that this also happens when pressing it on a filled Phial or Jar directly, if that's desirable (would mirror the behavior for fluid ex. water bucket).
> 
> I don't have an answer to that. I don't play with magic and I don't know how it would be convenient.

Well aspects aren't actual items, so you would have to go out of your way to search them in NEI before you could ever make use of this feature.

Expanding this feature to include phials and jars sounds like a sensible way to allow people to discover it. Most likely you would want to interpret the search as an "item search" if the phial/jar is empty and as an "aspect search" otherwise.

**Comment by leumasme** [comment]
> It is impossible to build a filter from the side of the ItemStack you are looking for, since the filter is unique for each TileEntity/ItemStack in which you need to search (If you have an idea how to fix this, give an example)
 
It's possible - just do the logic of extracting the needed data from the TE in the match method of the filter instead of the constructor. The Filter would then be specific to the item we're searching for instead of the TE/Stack in which we're searching, and the logic to process the inspected TE/Stack would be moved from the constructor of the filter to the match method.
In pseudocode, currently it's
```
class SomeFilter {
constructor(inspectedThing) {
  // Extract needed data from inspected thing, ex. for backpack/chest, get the contents of it
}
boolean match(searchedItem/searchRequest) {
  // Extract needed data from the Item we're searching for, ex. for water bucket, get the water fluid
  // Check if the extracted data of the inspectedThing matches the extractedData of the searchedItem
}
}
```
if we just invert it, then it's
```
class SomeFilter {
constructor(searchedItem/searchRequest) {
  // Extract needed data from the Item we're searching for, ex. for water bucket, get the water fluid
}
boolean match(inspectedThing) {
  // Extract needed data from inspected thing, ex. for backpack/chest, get the contents of it
  // Check if the extracted data of the inspectedThing matches the extractedData of the searchedItem
}
}
```
(Whether the extracting from inspectedThing is actually done in the constructor or in the getFilter which calls the constructor is irrelevant here, it's the same)
Since now the filter is created for the searched item instead of the inspected TE/Item, it can be re-used so we don't have to call the constructor so amny times and we don't have to repeat extracting the data from the searchedItem every time.

This would mean the sigature of `IStackFilter` would change from
```java
public interface IStackFilter {
    boolean matches(FindItemRequest request);
}
```
to
```java
public interface IStackFilter {
    boolean matches(ItemStack stack);
    boolean matches(TileEntity tileEntity);
}
```
and similarly, `getFilter` of providers would get passed the `FindItemRequest` or the searched `ItemStack` instead of the inspected `ItemStack`/`TileEntity`.
Now, for example for a Backpack Provider, the IStackFilter would be constructed with the item we're searching for, and the `matches(ItemStack stack)` method would check if `stack` is a backpack, extract the inventory from the backpack, and check if the inventory contains the stack that it was constructed with.
This does mean that most Providers won't return `null` for getFilter in most cases, but it would still drastically reduce the amount of Filters constructed, since we're now scaling with the amount of Providers instead of the amount of inspected Stacks/TEs

**Comment by leumasme** [comment]
https://github.com/user-attachments/assets/0ae9ad7a-eff6-4ba2-99a7-bd4cac7d11ef

Now with searching by essentia container

**Review by slprime — APPROVED** [review]

**Pull request merged**